tools/c7n_mailer/c7n_mailer/ldap_lookup.py

```diff
@@ -0,0 +1,174 @@
+        ldap_user_metadata = {attr.key: attr.value for attr in ldap_user_object}
+        ldap_user_metadata['dn'] = ldap_user_object.entry_dn
+        ldap_user_metadata[self.email_key] = ldap_user_metadata[self.email_key].lower()
+        ldap_user_metadata[self.uid_key] = ldap_user_metadata[self.uid_key].lower()
+        return ldap_user_metadata
+
+    # eg, uid = bill_lumbergh
+    def get_metadata_from_uid(self, uid):
+        uid = uid.lower()
+        if self.uid_regex:
+            # for example if you set ldap_uid_regex in your mailer.yml to "^[0-9]{6}$" then it
+            # would only query LDAP if your string length is 6 characters long and only digits.
+            # re.search("^[0-9]{6}$", "123456")
+            # Out[41]: <_sre.SRE_Match at 0x1109ab440>
+            # re.search("^[0-9]{6}$", "1234567") returns None, or "12345a' also returns None
+            if not re.search(self.uid_regex, uid):
+                regex_msg = 'uid does not match regex: %s %s' % (self.uid_regex, uid)
+                self.log.debug(regex_msg)
+                return {}
+        if self.cache_engine:
+            cache_result = self.caching.get(uid)
+            if cache_result or cache_result == {}:
+                cache_msg = 'Got ldap metadata from local cache for: %s' % uid
+                self.log.debug(cache_msg)
+                return cache_result
+        ldap_filter = '(%s=%s)' % (self.uid_key, uid)
+        ldap_results = self.search_ldap(self.base_dn, ldap_filter, attributes=self.attributes)
+        if ldap_results:
+            ldap_user_metadata = self.get_dict_from_ldap_object(self.connection.entries[0])
+        else:
+            self.caching.set(uid, {})
+            return {}
+        if self.cache_engine:
+            self.log.debug('Writing user: %s metadata to cache engine.' % uid)
+            self.caching.set(ldap_user_metadata['dn'], ldap_user_metadata)
+            self.caching.set(uid, ldap_user_metadata)
+        return ldap_user_metadata
+
+
+# the purpose of this class is to create a light wrapper around a standard library called shelve
+# which can store key/value pairs to a local file (using pickle), similar to how memcache or redis
+# stores things. This normalizes the methods to set/get a function, so you can interchangeable
+# decide which caching system to use, a local file, or memcache, redis, etc
+# If you don't want a redis dependency and aren't running the mailer in lambda this works well
+class LocalShelf(object):
+    def __init__(self, local_filename):
+        self.shelve = shelve.open(local_filename, flag='c')
+
+    def get(self, key):
+        return self.shelve.get(str(key))
+
+    def set(self, key, value):
+        self.shelve[key] = value
+
+
+# redis can't write complex python objects like dictionaries as values (the way memcache can)
+# so we turn our dict into a json string when setting, and json.loads when getting
+class Redis(object):
```

Comment:
It turns out redis can't write python dict objects like memcache can, it has to take strings. So this wrapper class is necessary to normalize the gets/gets. I just tested this flow with redis on my mac, works fine.